v2/svid/x509svid/svid.go

```diff
@@ -136,41 +118,128 @@ func (s *SVID) GetX509SVID() (*SVID, error) {
 	return s, nil
 }
 
-func newSVID(certificates []*x509.Certificate, privateKey interface{}) (*SVID, error) {
+func newSVID(certificates []*x509.Certificate, privateKey crypto.PrivateKey) (*SVID, error) {
+	spiffeID, err := validateCertificates(certificates)
+	if err != nil {
+		return nil, x509svidErr.New("certificate validation failed: %v", err)
+	}
+
+	signer, err := validatePrivateKey(privateKey, certificates[0])
+	if err != nil {
+		return nil, x509svidErr.New("private key validation failed: %v", err)
+	}
+
+	return &SVID{
+		Certificates: certificates,
+		PrivateKey:   signer,
+		ID:           *spiffeID,
+	}, nil
+}
+
+// validate the slice of certificates constitutes a valid SVID chain according
+// to the spiffe standard and returns the spiffe id of the leaf certificate
+func validateCertificates(certificates []*x509.Certificate) (*spiffeid.ID, error) {
 	if len(certificates) == 0 {
-		return nil, x509SVIDErr.New("no certificates found")
+		return nil, errs.New("no certificates found")
+	}
+
+	leafID, err := validateLeafCertificate(certificates[0])
+	if err != nil {
+		return nil, err
+	}
+
+	if len(certificates) == 1 {
```

Comment:
nit: how would you feel about removing this if statement?  `certificates[1:]` will be an empty slice in this situation and `validateSigningCertificates` will succeed since there is nothing to validate.

Reply:
yea! I totally agree, I'll remove it